### PR TITLE
Improve ValidateCount attribute error message

### LINKED ISSUE #3585: ValidateCount parameter-validation attribute error messages are misleading and inflexible
Steps to reproduce
------------------

```powershell
function foo { param([ValidateCount(2,2)] [string[]] $bar) }; foo 1; foo 1, 2, 3
```

Expected behavior
-----------------

An error message indicating that _exactly 2_ arguments (array elements) are expected.

Actual behavior
---------------

```
Cannot validate argument on parameter 'bar'. The number of provided arguments (1) is fewer than the minimum number of allowed arguments (2). Provide more than 2 arguments, 
and then try the command again.

Cannot validate argument on parameter 'bar'. The number of provided arguments, (3), exceeds the maximum number of allowed arguments (2). Provide fewer than 2 arguments, and 
then try the command again.
```

Note:

* The error messages do not take into account that the lower and upper bound are the same.

* Aside from that, the _wording is incorrect_: it should say "2 or more" rather than "more than 2" and "2 or fewer" rather than "fewer than 2".

Environment data
----------------

```powershell
PowerShell Core v6.0.0-alpha (v6.0.0-alpha.18) on Darwin Kernel Version 16.5.0: Fri Mar  3 16:52:33 PST 2017; root:xnu-3789.51.2~3/RELEASE_X86_64
```

### PR DESCRIPTION
**Pull request by iSazonov**
Fix #3585 

- [x] - Improve ValidateCount attribute error message
- [x] - Add tests (blocked by #3456)

### CODE REVIEW

**Review by BrucePay — DISMISSED** [review]
There's a typo in the error message. It should say "is not equal" - the "is" is missing.

**Comment by iSazonov** [comment]
Thanks! Typo is fixed.

File: src/System.Management.Automation/resources/Metadata.resx [hunk 1]
```diff
@@ -150,6 +150,9 @@
   <data name="ValidateCountMaxLengthFailure" xml:space="preserve">
     <value>The number of provided arguments, ({1}), exceeds the maximum number of allowed arguments ({0}). Provide fewer than {0} arguments, and then try the command again.</value>
   </data>
+  <data name="ValidateCountEqualLengthFailure" xml:space="preserve">
+    <value>The number of provided arguments, ({1}), is not equal the number of allowed arguments ({0}). Provide exactly {0} arguments, and then try the command again.</value>
```
Comment (lzybkr) [inline src/System.Management.Automation/resources/Metadata.resx]: This is definitely better, but I'm finding all 3 error messages too wordy and slightly inaccurate. First, maybe we should change `arguments` to `values` because there is a single argument with multiple values.

Second, maybe 1 error message is sufficient if we change it to:

```
The parameter requires at least `{0}` value(s) and no more than `{1}` value(s) - `{2}` value(s) were provided.
```

**Comment by daxian-dbw** [comment]
I edited the changes to correct a typo and remove duplicate test cases.
@iSazonov please see @lzybkr's comment, which I totally agree.

**Comment by iSazonov** [comment]
Thanks! I agree that this new message is more clear.

File: src/System.Management.Automation/resources/Metadata.resx [hunk 2]
```diff
@@ -145,10 +145,13 @@
...
   <data name="ValidateCountMinLengthFailure" xml:space="preserve">
-    <value>The number of provided arguments ({1}) is fewer than the minimum number of allowed arguments ({0}). Provide more than {0} arguments, and then try the command again.</value>
+    <value>The parameter requires at least `{0}` value(s) and no more than `{1}` value(s) - `{2}` value(s) were provided.</value>
   </data>
   <data name="ValidateCountMaxLengthFailure" xml:space="preserve">
-    <value>The number of provided arguments, ({1}), exceeds the maximum number of allowed arguments ({0}). Provide fewer than {0} arguments, and then try the command again.</value>
+    <value>The parameter requires at least `{0}` value(s) and no more than `{1}` value(s) - `{2}` value(s) were provided.</value>
+  </data>
+  <data name="ValidateCountEqualLengthFailure" xml:space="preserve">
+    <value>The parameter requires at least `{0}` value(s) and no more than `{1}` value(s) - `{2}` value(s) were provided.</value>
```
Comment (daxian-dbw) [inline src/System.Management.Automation/resources/Metadata.resx]: You don't need 3 resource strings, one would be enough since they are the same. The resource string name could be `ValidateCountFailure`.

File: src/System.Management.Automation/engine/Attributes.cs [hunk 3]
```diff
@@ -1284,18 +1284,25 @@ protected override void Validate(object arguments, EngineIntrinsics engineIntrin
                     null, Metadata.ValidateCountNotInArray);
             }
 
+            if (MinLength == MaxLength && len != MaxLength)
+            {
+                throw new ValidationMetadataException("ValidateCountEqualLengthFailure",
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/Attributes.cs]: The error id `ValidateCountEqualLengthFailure` sounds confusing. Maybe `ValidateCountNotExactlyEqual`?

File: src/System.Management.Automation/resources/Metadata.resx (hunk 2 above)
Comment (iSazonov) [inline src/System.Management.Automation/resources/Metadata.resx]: I was very generous 😄 
Fixed.

File: src/System.Management.Automation/engine/Attributes.cs (hunk 3 above)
Comment (iSazonov) [inline src/System.Management.Automation/engine/Attributes.cs]: Fixed.

File: src/System.Management.Automation/resources/Metadata.resx [hunk 4]
```diff
@@ -144,11 +144,8 @@
   <data name="ValidateCountNotInArray" xml:space="preserve">
     <value>The ValidateCount attribute cannot be applied to a non-array parameter. Either remove the attribute from the parameter or make the parameter an array parameter.</value>
   </data>
-  <data name="ValidateCountMinLengthFailure" xml:space="preserve">
-    <value>The number of provided arguments ({1}) is fewer than the minimum number of allowed arguments ({0}). Provide more than {0} arguments, and then try the command again.</value>
-  </data>
-  <data name="ValidateCountMaxLengthFailure" xml:space="preserve">
-    <value>The number of provided arguments, ({1}), exceeds the maximum number of allowed arguments ({0}). Provide fewer than {0} arguments, and then try the command again.</value>
+  <data name="ValidateCountFailure" xml:space="preserve">
+    <value>The parameter requires at least `{0}` value(s) and no more than `{1}` value(s) - `{2}` value(s) were provided.</value>
```
Comment (daxian-dbw) [inline src/System.Management.Automation/resources/Metadata.resx]: We don't use backtick ` in error strings. I think you can just remove the backticks here.

File: test/powershell/engine/ValidateAttributes.Tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+Describe 'Validate Attributes Tests' -Tags 'CI' {
+
+    Context "ValidateCount" {
+        BeforeAll {
+           $testCases = @(
+               @{ sb = { function Local:foo { param([ValidateCount(-1,2)] [string[]] $bar) }; foo         }; FullyQualifiedErrorId = "ExceptionConstructingAttribute";             InnerErrorId = "" }
+               @{ sb = { function Local:foo { param([ValidateCount(1,-1)] [string[]] $bar) }; foo         }; FullyQualifiedErrorId = "ExceptionConstructingAttribute";             InnerErrorId = "" }
+               @{ sb = { function Local:foo { param([ValidateCount(2, 1)] [string[]] $bar) }; foo         }; FullyQualifiedErrorId = "ValidateRangeMaxLengthSmallerThanMinLength"; InnerErrorId = "" }
+               @{ sb = { function Local:foo { param([ValidateCount(2, 2)] [string[]] $bar) }; foo 1       }; FullyQualifiedErrorId = "ParameterArgumentValidationError,foo";       InnerErrorId = "ValidateCountEqualLengthFailure" }
```
Comment (daxian-dbw) [inline test/powershell/engine/ValidateAttributes.Tests.ps1]: Error id has changed to `ValidateCountNotExactlyEqual`

Comment (iSazonov) [inline test/powershell/engine/ValidateAttributes.Tests.ps1]: Fixed.

**Comment by mklement0** [comment]
It's great that you fixed this so quickly, but while the single error message is now _technically_ accurate, I fear it doesn't provide the best guidance to the _user_.

As a user who has just mistakenly provided the wrong number of values `x`, I would want the error message to _address my specific error_, for a range `a` to `z` (simplified wording to illustrate the logic):

* If `a` equals `z`: "You specified `x`, but you must specify _exactly_ `a`."



* Otherwise:

  * If `x` < `a`: "You specified `x`, but you must specify _at least_ `a` (and no more than `z`)."

  * If `x` > `z`: "You specified `x`, but you must specify _no more than_ `z` (but at least `a`)."

File: src/System.Management.Automation/engine/Attributes.cs [hunk 6]
```diff
@@ -1284,18 +1284,25 @@ protected override void Validate(object arguments, EngineIntrinsics engineIntrin
                     null, Metadata.ValidateCountNotInArray);
             }
 
+            if (MinLength == MaxLength && len != MaxLength)
```
Comment (lzybkr) [inline src/System.Management.Automation/engine/Attributes.cs]: Isn't this case unnecessary now?

In fact, with one error message, there should be a single throw statement.

Comment (iSazonov) [inline src/System.Management.Automation/engine/Attributes.cs]: We should to keep the two original fully qualified error ids to avoid a breaking change. But the new one I could to remove:

What is your conclusion about last @mklement0 [comment](https://github.com/PowerShell/PowerShell/pull/3596#issuecomment-296338493)?
If you reject it I shall only remove the ` if (MinLength == MaxLength && len != MaxLength)`.
If you accept please give me right template.

Comment (lzybkr) [inline src/System.Management.Automation/engine/Attributes.cs]: The fully qualified error ids on parameter validation errors aren't too interesting, I'm not worried about changing those.

Personal opinion - adding prescriptive guidance may have been the cause for the original bug - having a confusing error message. It's easier to get technically accurate errors - what is wrong - than it is to suggest how to fix the problem, so it can be better to not bother. And in this specific case, I think it's fairly obvious how to fix the error.

Comment (iSazonov) [inline src/System.Management.Automation/engine/Attributes.cs]: Thanks! If @daxian-dbw  and @mklement0 do not object I'll create new PR tomorrow to only remove unneeded resource strings.

Comment (mklement0) [inline src/System.Management.Automation/engine/Attributes.cs]: @lzybkr 
I think the fundamental question is: Is it a _design goal_ to provide prescriptive guidance? If it is (I think it should be), then let's _fix_ and improve the currently _broken_ / incomplete guidance - and that would mean fixing the 2 broken messages and adding one for the exact-count case.

There are limits to how helpful a _generic_ validation error message can be (see #3630), but here's an opportunity to be as helpful as possible.

Personally, if I specified an incorrect number of values to a parameter that accepts an exact number and I got the following message, I'd feel like I'm being presented with a brain-teaser rather than a helpful error message:
`The parameter requires at least 2 value(s) and no more than 2 value(s).`

Comment (lzybkr) [inline src/System.Management.Automation/engine/Attributes.cs]: @mklement0 - you bring up two different issues - providing clear messages and providing prescriptive guidance.

In this case, one generic message is a bit weird, but felt like a corner case to me. After a quick [search](https://github.com/search?p=1&q=language%3Apowershell+ValidateCount&type=Code&utf8=%E2%9C%93) - `ValidateCount` isn't used that much (lots of duplicate hits from forks in the 426 results), but there are examples where Min==Max, so it's not as much of a corner case as I originally thought.

So I guess have 2 messages seems reasonable, but 3 still feels excessive.

My original thinking was - less and simpler code is ideal, and fewer error messages are also good (e.g. it reduces the translation efforts.)

As for prescriptive guidance, it's definitely not a goal for all error messages. To apply that standard widely, you'd end up with messages like `Missing closing '}' in statement block or type definition, add the closing '}' and try again.` And that feels unnecessary.

Comment (mklement0) [inline src/System.Management.Automation/engine/Attributes.cs]: @lzybkr: You're right: "prescriptive" was not the right term (and I agree with not needing to go that far) - "specific to the error made" would have been better.

Comment (iSazonov) [inline src/System.Management.Automation/engine/Attributes.cs]: What is a conclusion for new PR - leave one message and remove unneeded resource strings?

Comment (mklement0) [inline src/System.Management.Automation/engine/Attributes.cs]: @iSazonov: @lzybkr will have to confirm, but here's my understanding:

* For the `MinLength != MaxLength` case: Use the just-added `The parameter requires at least {0} value(s) and no more than {1} value(s) - {2} value(s) were provided.` message.

* For the `MinLength == MaxLength` case: add a new message along the lines of `The parameter requires exactly {0} value(s)  - {2} value(s) were provided.`

Comment (lzybkr) [inline src/System.Management.Automation/engine/Attributes.cs]: @iSazonov @mklement0 - confirmed.
